doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,434 @@
+```go
+// Hooks of the User.
+func (User) Hooks() []ent.Hook {
+	return []ent.Hook{
+		hook.If(clearConnectionString,
+			hook.And(hook.Or(hook.HasOp(ent.OpUpdateOne), hook.HasOp(ent.OpCreate)), hook.HasFields(user.FieldConnectionString)),
+		),
+	}
+}
+```
+* Do not forget to run go generate or the hook will not be registered.
+
+A few hook helpers are used here, ```clearConnectionString``` is the name of the hook function, and it will run if a user 
+is Updated or Created and the mutation has the field ```connection_string``` (if we change only the name of the user there is no need to do anything right?).
+
+The hook:
+```go
+func clearConnectionString(next ent.Mutator) ent.Mutator {
+	return hook.UserFunc(func(ctx context.Context, m *gen.UserMutation) (ent.Value, error) {
+		cs, ok := m.ConnectionString()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		sp := strings.Split(cs, "@")
+		if len(sp) != 2 {
+			return next.Mutate(ctx, m)
+		}
+		sp = strings.Split(sp[0], ":")
+		if len(sp) != 3 {
+			return next.Mutate(ctx, m)
+		}
+		pass := sp[2]
+		m.SetPassword(pass)
+		m.SetConnectionString(strings.ReplaceAll(cs, pass, "****"))
+		return next.Mutate(ctx, m)
+	})
+} 
+```
+We validate that the mutation has a connection string (if not just let the mutation continue).
+A bit of string acrobatics, and we have the password extracted from the string, we store it in another field - the ```password``` and
+change the ```connection_string``` with a masked version. 
+
+To get the full connection string we add a utility function 
+```go
+package ent
+
+import "strings"
+
+// FullConnectionString returns the connection string with an unmasked password.
+func (u *User) FullConnectionString() string {
+	return strings.ReplaceAll(u.ConnectionString, "****", u.Password)
+}
+```
+
+BTW, you can see that it's OK to add files to the ent directory (not only in schema folder) since the ent generate command only appends/changes ent
+generated files.
+
+
+### Special Validators:
+
+Ent provides [validators](https://entgo.io/docs/schema-fields#validators) for fields, yet sometimes the logic can contain
+dependencies that are more applicative and require a broader context.
+For example, in our use case, we would like to make sure that dogs names don’t start with the first 2 letters of the owners name 
+(Best practice when you don't want your dog to run in circles every time someone calls your name).
+
+We start with a simple test
+```go
+func TestDogNameValidationHook(t *testing.T) {
+	ctx := context.Background()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+	)
+	u := c.User.Create().SetName("Yoni").
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	_, err := c.Dog.Create().SetName("Yolo").SetOwner(u).Save(ctx)
+	require.Error(t, err)
+}
+```
+
+In this case, you can see that since "Yoni" and "Yolo" share the same 2 letters "Yo" the action fails.
```

Comment:
Yolo is a great name though. At least for Kuro it would be :laughing:

Reply:
❤️